// meteor-desktop/dist/desktop.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _shelljs = _interopRequireDefault(require("shelljs"));

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _hashFiles = _interopRequireDefault(require("hash-files"));

var _log = _interopRequireDefault(require("./log"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

_shelljs.default.config.fatal = true;
/**
 * Checks if the path is empty.
 * @param {string} searchPath
 * @returns {boolean}
 */

function isEmptySync(searchPath) {
  var stat;

  try {
    stat = _fs.default.statSync(searchPath);
  } catch (e) {
    return true;
  }

  if (stat.isDirectory()) {
    var items = _fs.default.readdirSync(searchPath);

    return !items || !items.length;
  }

  return false;
}
/**
 * Represents the .desktop directory.
 * @class
 * @property {desktopSettings} settings
 */


var Desktop =
/*#__PURE__*/
function () {
  /**
   * @param {MeteorDesktop} $ - context
   *
   * @constructor
   */
  function Desktop($) {
    _classCallCheck(this, Desktop);

    this.$ = $;
    this.log = new _log.default('desktop');
    this.settings = null;
    this.dependencies = null;
  }
  /**
   * Tries to read and returns settings.json contents from .desktop dir.
   *
   * @returns {desktopSettings|null}
   */


  _createClass(Desktop, [{
    key: "getSettings",
    value: function getSettings() {
      if (!this.settings) {
        try {
          this.settings = JSON.parse(_fs.default.readFileSync(this.$.env.paths.desktop.settings, 'UTF-8'));
        } catch (e) {
          this.log.error('error while trying to read \'.desktop/settings.json\': ', e);
          process.exit(1);
        }
      }

      return this.settings;
    }
    /**
     * Returns a version hash representing current .desktop contents.
     * @returns {string}
     */

  }, {
    key: "getHashVersion",
    value: function getHashVersion() {
      this.log.info('calculating hash version from .desktop contents');

      var version = _hashFiles.default.sync({
        files: [`${this.$.env.paths.desktop.root}${_path.default.sep}**`]
      });

      this.log.verbose(`calculated .desktop hash version is ${version}`);
      return version;
    }
    /**
     * Tries to read a module.json file from a module at provided path.
     *
     * @param {string} modulePath - path to the module dir
     * @returns {Object}
     */

  }, {
    key: "getModuleConfig",
    value: function getModuleConfig(modulePath) {
      var moduleConfig = {};

      try {
        moduleConfig = JSON.parse(_fs.default.readFileSync(_path.default.join(modulePath, 'module.json'), 'UTF-8'));
      } catch (e) {
        this.log.error(`error while trying to read 'module.json' from '${modulePath}' module: `, e);
        process.exit(1);
      }

      if (!('name' in moduleConfig)) {
        this.log.error(`no 'name' field defined in 'module.json' in '${modulePath}' module.`);
        process.exit(1);
      }

      return moduleConfig;
    }
    /**
     * Scans all modules for module.json and gathers this configuration altogether.
     *
     * @returns {[]}
     */

  }, {
    key: "gatherModuleConfigs",
    value: function gatherModuleConfigs() {
      var _this = this;

      var configs = [];

      if (!isEmptySync(this.$.env.paths.desktop.modules)) {
        _shelljs.default.ls('-d', _path.default.join(this.$.env.paths.desktop.modules, '*')).forEach(function (module) {
          if (_fs.default.lstatSync(module).isDirectory()) {
            var moduleConfig = _this.getModuleConfig(module);

            moduleConfig.dirName = _path.default.parse(module).name;
            configs.push(moduleConfig);
          }
        });
      }

      return configs;
    }
    /**
     * Summarizes all dependencies defined in .desktop.
     *
     * @params {Object} settings      - settings.json
     * @params {boolean} checkModules - whether to gather modules dependencies
     * @params {boolean} refresh      - recompute
     * @returns {{fromSettings: {}, plugins: {}, modules: {}}}
     */

  }, {
    key: "getDependencies",
    value: function getDependencies() {
      var _this2 = this;

      var settings = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
      var checkModules = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
      var refresh = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;

      if (!refresh && this.dependencies) {
        return this.dependencies;
      }

      var dependencies = {
        fromSettings: {},
        plugins: {},
        modules: {}
      };
      /** @type {desktopSettings} * */

      var settingsJson = settings || this.getSettings(); // Settings can have a 'dependencies' field.

      if ('dependencies' in settingsJson) {
        dependencies.fromSettings = settingsJson.dependencies;
      } // Plugins are also a npm packages.


      if ('plugins' in settingsJson) {
        dependencies.plugins = Object.keys(settingsJson.plugins).reduce(function (plugins, plugin) {
          /* eslint-disable no-param-reassign */
          if (typeof settingsJson.plugins[plugin] === 'object') {
            plugins[plugin] = settingsJson.plugins[plugin].version;
          } else {
            plugins[plugin] = settingsJson.plugins[plugin];
          }

          return plugins;
        }, {});
      } // Each module can have its own dependencies defined.


      var moduleDependencies = {};

      if (checkModules) {
        var configs = this.gatherModuleConfigs();
        configs.forEach(function (moduleConfig) {
          if (!('dependencies' in moduleConfig)) {
            moduleConfig.dependencies = {};
          }

          if (moduleConfig.name in moduleDependencies) {
            _this2.log.error(`duplicate name '${moduleConfig.name}' in 'module.json' in ` + `'${moduleConfig.dirName}' - another module already registered the same name.`);

            process.exit(1);
          }

          moduleDependencies[moduleConfig.name] = moduleConfig.dependencies;
        });
      }

      dependencies.modules = moduleDependencies;
      this.dependencies = dependencies;
      return dependencies;
    }
    /**
     * Copies the .desktop scaffold into the meteor app dir.
     * Adds entry to .meteor/.gitignore.
     */

  }, {
    key: "scaffold",
    value: function scaffold() {
      this.log.info('creating .desktop scaffold in your project');

      if (this.$.utils.exists(this.$.env.paths.desktop.root)) {
        this.log.warn('.desktop already exists - delete it if you want a new one to be ' + 'created');
        return;
      }

      _shelljs.default.cp('-r', this.$.env.paths.scaffold, this.$.env.paths.desktop.root);

      _shelljs.default.mkdir(this.$.env.paths.desktop.import);

      this.log.info('.desktop directory prepared');
    }
    /**
     * Verifies if all mandatory files are present in the .desktop.
     *
     * @returns {boolean}
     */

  }, {
    key: "check",
    value: function check() {
      this.log.verbose('checking .desktop existence');
      return !!(this.$.utils.exists(this.$.env.paths.desktop.root) && this.$.utils.exists(this.$.env.paths.desktop.settings) && this.$.utils.exists(this.$.env.paths.desktop.desktop));
    }
  }]);

  return Desktop;
}();
/**
 * @typedef {Object} desktopSettings
 * @property {string} name
 * @property {string} projectName
 * @property {boolean} devTools
 * @property {boolean} devtron
 * @property {boolean} desktopHCP
 * @property {string} autoUpdateFeedUrl
 * @property {Object} autoUpdateFeedHeaders
 * @property {Object} autoUpdateManualCheck
 * @property {Object} desktopHCPSettings
 * @property {boolean} desktopHCPSettings.ignoreCompatibilityVersion
 * @property {boolean} desktopHCPSettings.blockAppUpdateOnDesktopIncompatibility
 * @property {number} webAppStartupTimeout
 * @property {Object} window
 * @property {Object} windowDev
 * @property {Object} packageJsonFields
 * @property {Object} builderOptions
 * @property {Object} builderCliOptions
 * @property {Object} packagerOptions
 * @property {Object} plugins
 * @property {Object} dependencies
 * @property {boolean} uglify
 * @property {string} version
 * */


exports.default = Desktop;
