// meteor-desktop/dist/log.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/* eslint-disable no-console */

/*
 0.OFF
 1.INFO
 2.WARN
 3.ERROR
 4.TRACE
 5.DEBUG
 6.ALL
 */
var Log =
/*#__PURE__*/
function () {
  function Log(prefix) {
    _classCallCheck(this, Log);

    this.prefix = prefix;
  }

  _createClass(Log, [{
    key: "log",
    value: function log(type, args) {
      console.log.apply(null, [`${type}  ${this.prefix}: `].concat(Log.slice(args)));
    }
  }, {
    key: "info",
    value: function info() {
      if (/INFO|ALL/i.test(Log.level())) {
        for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }

        this.log('INFO', args);
      }
    }
  }, {
    key: "warn",
    value: function warn() {
      if (/WARN|ALL/i.test(Log.level())) {
        for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
          args[_key2] = arguments[_key2];
        }

        this.log('WARN', args);
      }
    }
  }, {
    key: "error",
    value: function error() {
      if (/ERROR|ALL/i.test(Log.level())) {
        for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
          args[_key3] = arguments[_key3];
        }

        this.log('ERROR', args);
      }
    }
  }, {
    key: "debug",
    value: function debug() {
      if (/DEBUG|ALL/i.test(Log.level())) {
        for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
          args[_key4] = arguments[_key4];
        }

        this.log('DEBUG', args);
      }
    }
  }, {
    key: "verbose",
    value: function verbose() {
      if (/VERBOSE|ALL/i.test(Log.level())) {
        for (var _len5 = arguments.length, args = new Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
          args[_key5] = arguments[_key5];
        }

        this.log('VERBOSE', args);
      }
    }
  }, {
    key: "trace",
    value: function trace() {
      if (/TRACE|ALL/i.test(Log.level())) {
        for (var _len6 = arguments.length, args = new Array(_len6), _key6 = 0; _key6 < _len6; _key6++) {
          args[_key6] = arguments[_key6];
        }

        this.log('TRACE', args);
      }
    }
  }], [{
    key: "level",
    value: function level() {
      return process.env.MD_LOG_LEVEL || 'INFO,WARN,ERROR';
    }
  }, {
    key: "slice",
    value: function slice(args) {
      return Array.prototype.slice.call(args, 0);
    }
  }]);

  return Log;
}();

exports.default = Log;
