// meteor-desktop/dist/meteorManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _crossSpawn = _interopRequireDefault(require("cross-spawn"));

var _log3 = _interopRequireDefault(require("./log"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } } function _next(value) { step("next", value); } function _throw(err) { step("throw", err); } _next(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

/**
 * Utility class designed for managing Meteor packages.
 *
 * @property {MeteorDesktop} $
 * @class
 */
var MeteorManager =
/*#__PURE__*/
function () {
  /**
   * @param {MeteorDesktop} $ - context
   * @constructor
   */
  function MeteorManager($) {
    _classCallCheck(this, MeteorManager);

    this.log = new _log3.default('meteorManager');
    this.$ = $;
  }
  /**
   * Looks for specified packages in .meteor/packages. In other words checks if the project has
   * specified packages added.
   * @param {Array} packages
   * @returns {boolean}
   */


  _createClass(MeteorManager, [{
    key: "checkPackages",
    value: function checkPackages(packages) {
      var usedPackages = _fs.default.readFileSync(this.$.env.paths.meteorApp.packages, 'UTF-8').replace(/\r/gm, '').split('\n').filter(function (line) {
        return !line.trim().startsWith('#');
      });

      return !packages.some(function (packageToFind) {
        return !usedPackages.some(function (meteorPackage) {
          return ~meteorPackage.indexOf(packageToFind);
        });
      });
    }
    /**
     * Looks for specified packages in .meteor/packages. In other words checks if the project has
     * specified packages added.
     * @param {Array} packages
     * @returns {boolean}
     */

  }, {
    key: "checkPackagesVersion",
    value: function checkPackagesVersion(packages) {
      var usedPackages = _fs.default.readFileSync(this.$.env.paths.meteorApp.versions, 'UTF-8').replace(/\r/gm, '').split('\n');

      return !packages.some(function (packageToFind) {
        return !usedPackages.some(function (meteorPackage) {
          return meteorPackage === packageToFind;
        });
      });
    }
    /**
     * Ensures certain packages are added to meteor project and in correct version.
     * @param {Array} packages
     * @param {Array} packagesWithVersion
     * @param {string} who - name of the entity that requests presence of thos packages (can be the
     *                       integration itself or a plugin)
     * @returns {Promise.<void>}
     */

  }, {
    key: "ensurePackages",
    value: function () {
      var _ensurePackages = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee(packages, packagesWithVersion, who) {
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                if (this.checkPackages(packages)) {
                  _context.next = 10;
                  break;
                }

                this.log.warn(`${who} requires some packages that are not added to project, will try to add them now`);
                _context.prev = 2;
                _context.next = 5;
                return this.addPackages(packages, packagesWithVersion);

              case 5:
                _context.next = 10;
                break;

              case 7:
                _context.prev = 7;
                _context.t0 = _context["catch"](2);
                throw new Error(_context.t0);

              case 10:
                if (this.checkPackagesVersion(packagesWithVersion)) {
                  _context.next = 20;
                  break;
                }

                this.log.warn(`${who} required packages version is different, fixing it`);
                _context.prev = 12;
                _context.next = 15;
                return this.addPackages(packages, packagesWithVersion);

              case 15:
                _context.next = 20;
                break;

              case 17:
                _context.prev = 17;
                _context.t1 = _context["catch"](12);
                throw new Error(_context.t1);

              case 20:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[2, 7], [12, 17]]);
      }));

      return function ensurePackages(_x, _x2, _x3) {
        return _ensurePackages.apply(this, arguments);
      };
    }()
    /**
     * Removes packages from the meteor app.
     * @param {Array} packages            - array with names of the packages to remove
     */

  }, {
    key: "deletePackages",
    value: function deletePackages(packages) {
      var _log,
          _this = this;

      (_log = this.log).warn.apply(_log, ['removing packages from meteor project'].concat(_toConsumableArray(packages)));

      return new Promise(function (resolve, reject) {
        (0, _crossSpawn.default)('meteor', ['remove'].concat(packages), {
          cwd: _this.$.env.paths.meteorApp.root,
          stdio: ['pipe', 'pipe', process.stderr],
          env: Object.assign({
            METEOR_PRETTY_OUTPUT: 0,
            METEOR_NO_RELEASE_CHECK: 1
          }, process.env)
        }).on('exit', function (code) {
          if (code !== 0 || _this.checkPackages(packages)) {
            reject('removeing packages failed');
          } else {
            resolve();
          }
        });
      });
    }
    /**
     * Adds packages to the meteor app.
     * @param {Array} packages            - array with names of the packages to add
     * @param {Array} packagesWithVersion - array with names and versions of the packages to add
     */

  }, {
    key: "addPackages",
    value: function addPackages(packages, packagesWithVersion) {
      var _log2,
          _this2 = this;

      (_log2 = this.log).info.apply(_log2, ['adding packages to meteor project'].concat(_toConsumableArray(packagesWithVersion)));

      return new Promise(function (resolve, reject) {
        (0, _crossSpawn.default)('meteor', ['add'].concat(packagesWithVersion.map(function (packageName) {
          return packageName.replace('@', '@=');
        })), {
          cwd: _this2.$.env.paths.meteorApp.root,
          stdio: ['pipe', 'pipe', process.stderr],
          env: Object.assign({
            METEOR_PRETTY_OUTPUT: 0,
            METEOR_NO_RELEASE_CHECK: 1
          }, process.env)
        }).on('exit', function (code) {
          if (code !== 0 || !_this2.checkPackages(packages)) {
            reject('adding packages failed');
          } else {
            resolve();
          }
        });
      });
    }
  }]);

  return MeteorManager;
}();

exports.default = MeteorManager;
