// meteor-desktop/dist/meteorApp.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _crossSpawn = _interopRequireDefault(require("cross-spawn"));

var _semver = _interopRequireDefault(require("semver"));

var _shelljs = _interopRequireDefault(require("shelljs"));

var _path = _interopRequireDefault(require("path"));

var _singleLineLog = _interopRequireDefault(require("single-line-log"));

var _asar = _interopRequireDefault(require("asar"));

var _nodeFetch = _interopRequireDefault(require("node-fetch"));

var _isDesktopInjector = _interopRequireDefault(require("../skeleton/modules/autoupdate/isDesktopInjector"));

var _log = _interopRequireDefault(require("./log"));

var _meteorManager = _interopRequireDefault(require("./meteorManager"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _sliceIterator(arr, i) { var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _slicedToArray(arr, i) { if (Array.isArray(arr)) { return arr; } else if (Symbol.iterator in Object(arr)) { return _sliceIterator(arr, i); } else { throw new TypeError("Invalid attempt to destructure non-iterable instance"); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } } function _next(value) { step("next", value); } function _throw(err) { step("throw", err); } _next(); }); }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

var join = _path.default.join;
var sll = _singleLineLog.default.stdout; // TODO: refactor all strategy ifs to one place

/**
 * Represents the Meteor app.
 * @property {MeteorDesktop} $
 * @class
 */

var MeteorApp =
/*#__PURE__*/
function () {
  /**
   * @param {MeteorDesktop} $ - context
   * @constructor
   */
  function MeteorApp($) {
    _classCallCheck(this, MeteorApp);

    this.log = new _log.default('meteorApp');
    this.$ = $;
    this.meteorManager = new _meteorManager.default($);
    this.mobilePlatform = null;
    this.oldManifest = null;
    this.injector = new _isDesktopInjector.default();
    this.matcher = new RegExp('__meteor_runtime_config__ = JSON.parse\\(decodeURIComponent\\("([^"]*)"\\)\\)');
    this.replacer = new RegExp('(__meteor_runtime_config__ = JSON.parse\\(decodeURIComponent\\()"([^"]*)"(\\)\\))');
    this.meteorVersion = null;
    this.indexHTMLstrategy = null;
    this.indexHTMLStrategies = {
      INDEX_FROM_CORDOVA_BUILD: 1,
      INDEX_FROM_RUNNING_SERVER: 2
    };
  }
  /**
   * Ensures that required packages are added to the Meteor app.
   */


  _createClass(MeteorApp, [{
    key: "ensureDesktopHCPPackages",
    value: function () {
      var _ensureDesktopHCPPackages = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee() {
        var _this = this;

        var desktopHCPPackages, packagesWithVersion;
        return regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                desktopHCPPackages = ['omega:meteor-desktop-watcher', 'omega:meteor-desktop-bundler'];

                if (!this.$.desktop.getSettings().desktopHCP) {
                  _context.next = 14;
                  break;
                }

                this.log.verbose('desktopHCP is enabled, checking for required packages');
                packagesWithVersion = desktopHCPPackages.map(function (packageName) {
                  return `${packageName}@${_this.$.getVersion()}`;
                });
                _context.prev = 4;
                _context.next = 7;
                return this.meteorManager.ensurePackages(desktopHCPPackages, packagesWithVersion, 'desktopHCP');

              case 7:
                _context.next = 12;
                break;

              case 9:
                _context.prev = 9;
                _context.t0 = _context["catch"](4);
                throw new Error(_context.t0);

              case 12:
                _context.next = 24;
                break;

              case 14:
                this.log.verbose('desktopHCP is not enabled, removing required packages');
                _context.prev = 15;

                if (!this.meteorManager.checkPackages(desktopHCPPackages)) {
                  _context.next = 19;
                  break;
                }

                _context.next = 19;
                return this.meteorManager.deletePackages(desktopHCPPackages, 'desktopHCP');

              case 19:
                _context.next = 24;
                break;

              case 21:
                _context.prev = 21;
                _context.t1 = _context["catch"](15);
                throw new Error(_context.t1);

              case 24:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this, [[4, 9], [15, 21]]);
      }));

      return function ensureDesktopHCPPackages() {
        return _ensureDesktopHCPPackages.apply(this, arguments);
      };
    }()
    /**
     * Adds entry to .meteor/.gitignore if necessary.
     */

  }, {
    key: "updateGitIgnore",
    value: function updateGitIgnore() {
      this.log.verbose('updating .meteor/.gitignore'); // Lets read the .meteor/.gitignore and filter out blank lines.

      var gitIgnore = _fs.default.readFileSync(this.$.env.paths.meteorApp.gitIgnore, 'UTF-8').split('\n').filter(function (ignoredPath) {
        return ignoredPath.trim() !== '';
      });

      if (!~gitIgnore.indexOf(this.$.env.paths.electronApp.rootName)) {
        this.log.verbose(`adding ${this.$.env.paths.electronApp.rootName} to .meteor/.gitignore`);
        gitIgnore.push(this.$.env.paths.electronApp.rootName);

        _fs.default.writeFileSync(this.$.env.paths.meteorApp.gitIgnore, gitIgnore.join('\n'), 'UTF-8');
      }
    }
    /**
     * Reads the Meteor release version used in the app.
     * @returns {string}
     */

  }, {
    key: "getMeteorRelease",
    value: function getMeteorRelease() {
      var release = _fs.default.readFileSync(this.$.env.paths.meteorApp.release, 'UTF-8').replace(/\r/gm, '').split('\n')[0];

      var _release$split = release.split('@');

      var _release$split2 = _slicedToArray(_release$split, 2);

      release = _release$split2[1];

      // We do not care if it is beta.
      if (~release.indexOf('-')) {
        var _release$split3 = release.split('-');

        var _release$split4 = _slicedToArray(_release$split3, 1);

        release = _release$split4[0];
      }

      return release;
    }
    /**
     * Cast Meteor release to semver version.
     * @returns {string}
     */

  }, {
    key: "castMeteorReleaseToSemver",
    value: function castMeteorReleaseToSemver() {
      return `${this.getMeteorRelease()}.0.0`.match(/(^\d+\.\d+\.\d+)/gmi)[0];
    }
    /**
     * Validate meteor version against a versionRange.
     * @param {string} versionRange - semver version range
     */

  }, {
    key: "checkMeteorVersion",
    value: function checkMeteorVersion(versionRange) {
      var release = this.castMeteorReleaseToSemver();

      if (!_semver.default.satisfies(release, versionRange)) {
        if (this.$.env.options.skipMobileBuild) {
          this.log.error(`wrong meteor version (${release}) in project - only ` + `${versionRange} is supported`);
        } else {
          this.log.error(`wrong meteor version (${release}) in project - only ` + `${versionRange} is supported for automatic meteor builds (you can always ` + 'try with `--skip-mobile-build` if you are using meteor >= 1.2.1');
        }

        process.exit(1);
      }
    }
    /**
     * Decides which strategy to use while trying to get client build out of Meteor project.
     * @returns {number}
     */

  }, {
    key: "chooseStrategy",
    value: function chooseStrategy() {
      if (this.$.env.options.forceCordovaBuild) {
        return this.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD;
      }

      var release = this.castMeteorReleaseToSemver();

      if (_semver.default.satisfies(release, '> 1.3.4')) {
        return this.indexHTMLStrategies.INDEX_FROM_RUNNING_SERVER;
      }

      if (_semver.default.satisfies(release, '1.3.4')) {
        var explodedVersion = this.getMeteorRelease().split('.');

        if (explodedVersion.length >= 4) {
          if (explodedVersion[3] > 1) {
            return this.indexHTMLStrategies.INDEX_FROM_RUNNING_SERVER;
          }

          return this.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD;
        }
      }

      return this.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD;
    }
    /**
     * Checks required preconditions.
     * - Meteor version
     * - is mobile platform added
     */

  }, {
    key: "checkPreconditions",
    value: function () {
      var _checkPreconditions = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee2() {
        var platforms;
        return regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                if (this.$.env.options.skipMobileBuild) {
                  this.checkMeteorVersion('>= 1.2.1');
                } else {
                  this.checkMeteorVersion('>= 1.3.3');
                  this.indexHTMLstrategy = this.chooseStrategy();

                  if (this.indexHTMLstrategy === this.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD) {
                    this.log.debug('meteor version is < 1.3.4.2 so the index.html from cordova-build will' + ' be used');
                  } else {
                    this.log.debug('meteor version is >= 1.3.4.2 so the index.html will be downloaded ' + 'from __cordova/index.html');
                  }
                }

                if (this.$.env.options.skipMobileBuild) {
                  _context2.next = 15;
                  break;
                }

                platforms = _fs.default.readFileSync(this.$.env.paths.meteorApp.platforms, 'UTF-8');

                if (!(!~platforms.indexOf('android') && !~platforms.indexOf('ios'))) {
                  _context2.next = 15;
                  break;
                }

                if (!this.$.env.options.android) {
                  this.mobilePlatform = 'ios';
                } else {
                  this.mobilePlatform = 'android';
                }

                this.log.warn(`no mobile target detected - will add '${this.mobilePlatform}' ` + 'just to get a mobile build');
                _context2.prev = 6;
                _context2.next = 9;
                return this.addMobilePlatform(this.mobilePlatform);

              case 9:
                _context2.next = 15;
                break;

              case 11:
                _context2.prev = 11;
                _context2.t0 = _context2["catch"](6);
                this.log.error('failed to add a mobile platform - please try to do it manually');
                process.exit(1);

              case 15:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this, [[6, 11]]);
      }));

      return function checkPreconditions() {
        return _checkPreconditions.apply(this, arguments);
      };
    }()
    /**
     * Tries to add a mobile platform to meteor project.
     * @param {string} platform - platform to add
     * @returns {Promise}
     */

  }, {
    key: "addMobilePlatform",
    value: function addMobilePlatform(platform) {
      var _this2 = this;

      return new Promise(function (resolve, reject) {
        _this2.log.verbose(`adding mobile platform: ${platform}`);

        (0, _crossSpawn.default)('meteor', ['add-platform', platform], {
          cwd: _this2.$.env.paths.meteorApp.root,
          stdio: _this2.$.env.stdio
        }).on('exit', function () {
          var platforms = _fs.default.readFileSync(_this2.$.env.paths.meteorApp.platforms, 'UTF-8');

          if (!~platforms.indexOf('android') && !~platforms.indexOf('ios')) {
            reject();
          } else {
            resolve();
          }
        });
      });
    }
    /**
     * Tries to remove a mobile platform from meteor project.
     * @param {string} platform - platform to remove
     * @returns {Promise}
     */

  }, {
    key: "removeMobilePlatform",
    value: function removeMobilePlatform(platform) {
      var _this3 = this;

      return new Promise(function (resolve, reject) {
        _this3.log.verbose(`removing mobile platform: ${platform}`);

        (0, _crossSpawn.default)('meteor', ['remove-platform', platform], {
          cwd: _this3.$.env.paths.meteorApp.root,
          stdio: _this3.$.env.stdio,
          env: Object.assign({
            METEOR_PRETTY_OUTPUT: 0
          }, process.env)
        }).on('exit', function () {
          var platforms = _fs.default.readFileSync(_this3.$.env.paths.meteorApp.platforms, 'UTF-8');

          if (~platforms.indexOf(platform)) {
            reject();
          } else {
            resolve();
          }
        });
      });
    }
    /**
     * Just checks for index.html and program.json existence.
     * @returns {boolean}
     */

  }, {
    key: "isCordovaBuildReady",
    value: function isCordovaBuildReady() {
      if (this.indexHTMLstrategy === this.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD) {
        return this.$.utils.exists(this.$.env.paths.meteorApp.cordovaBuildIndex) && this.$.utils.exists(this.$.env.paths.meteorApp.cordovaBuildProgramJson) && (!this.oldManifest || this.oldManifest && this.oldManifest !== _fs.default.readFileSync(this.$.env.paths.meteorApp.cordovaBuildProgramJson, 'UTF-8'));
      }

      return this.$.utils.exists(this.$.env.paths.meteorApp.webCordovaProgramJson) && (!this.oldManifest || this.oldManifest && this.oldManifest !== _fs.default.readFileSync(this.$.env.paths.meteorApp.webCordovaProgramJson, 'UTF-8'));
    }
    /**
     * Fetches index.html from running project.
     * @returns {Promise.<*>}
     */

  }, {
    key: "acquireIndex",
    value: function () {
      var _acquireIndex = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee3() {
        var port, res, text;
        return regeneratorRuntime.wrap(function _callee3$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                port = this.$.env.options.port ? this.$.env.options.port : 3080;
                this.log.info('acquiring index.html');
                _context3.next = 4;
                return (0, _nodeFetch.default)(`http://127.0.0.1:${port}/__cordova/index.html`);

              case 4:
                res = _context3.sent;
                _context3.next = 7;
                return res.text();

              case 7:
                text = _context3.sent;

                if (!~text.indexOf('src="/cordova.js"')) {
                  _context3.next = 10;
                  break;
                }

                return _context3.abrupt("return", text);

              case 10:
                return _context3.abrupt("return", false);

              case 11:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee3, this);
      }));

      return function acquireIndex() {
        return _acquireIndex.apply(this, arguments);
      };
    }()
    /**
     * Fetches mainfest.json from running project.
     * @returns {Promise.<void>}
     */

  }, {
    key: "acquireManifest",
    value: function () {
      var _acquireManifest = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee4() {
        var port, res, text;
        return regeneratorRuntime.wrap(function _callee4$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                port = this.$.env.options.port ? this.$.env.options.port : 3080;
                this.log.info('acquiring manifest.json');
                _context4.next = 4;
                return (0, _nodeFetch.default)(`http://127.0.0.1:${port}/__cordova/manifest.json?meteor_dont_serve_index=true`);

              case 4:
                res = _context4.sent;
                _context4.next = 7;
                return res.text();

              case 7:
                text = _context4.sent;
                return _context4.abrupt("return", JSON.parse(text));

              case 9:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee4, this);
      }));

      return function acquireManifest() {
        return _acquireManifest.apply(this, arguments);
      };
    }()
    /**
     * Tries to get a mobile build from meteor app.
     * In case of failure leaves a meteor.log.
     * A lot of stuff is happening here - but the main aim is to get a mobile build from
     * .meteor/local/cordova-build/www/application and exit as soon as possible.
     *
     * @returns {Promise}
     */

  }, {
    key: "buildMobileTarget",
    value: function buildMobileTarget() {
      var _this4 = this;

      var programJson = this.indexHTMLstrategy === this.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD ? this.$.env.paths.meteorApp.cordovaBuildProgramJson : this.$.env.paths.meteorApp.webCordovaProgramJson;

      if (this.$.utils.exists(programJson)) {
        this.oldManifest = _fs.default.readFileSync(programJson, 'UTF-8');
      }

      return new Promise(function (resolve, reject) {
        var self = _this4;
        var log = '';
        var desiredExit = false;
        var buildTimeout = null;
        var errorTimeout = null;
        var messageTimeout = null;
        var killTimeout = null;
        var cordovaCheckInterval = null;
        var portProblem = false;

        function windowsKill(pid) {
          self.log.debug(`killing pid: ${pid}`);

          _crossSpawn.default.sync('taskkill', ['/pid', pid, '/f', '/t']); // We will look for other process which might have been created outside the
          // process tree.
          // Lets list all node.exe processes.


          var out = _crossSpawn.default.sync('wmic', ['process', 'where', 'caption="node.exe"', 'get', 'commandline,processid']).stdout.toString('utf-8').split('\n');

          var args = self.prepareArguments(); // Lets mount regex.

          var regexV1 = new RegExp(`${args.join('\\s+')}\\s+(\\d+)`, 'gm');
          var regexV2 = new RegExp(`"${args.join('"\\s+"')}"\\s+(\\d+)`, 'gm'); // No we will check for those with the matching params.

          out.forEach(function (line) {
            var match = regexV1.exec(line) || regexV2.exec(line) || false;

            if (match) {
              self.log.debug(`killing pid: ${match[1]}`);

              _crossSpawn.default.sync('taskkill', ['/pid', match[1], '/f', '/t']);
            }

            regexV1.lastIndex = 0;
            regexV2.lastIndex = 0;
          });
        }

        function writeLog() {
          _fs.default.writeFileSync('meteor.log', log, 'UTF-8');
        }

        function clearTimeoutsAndIntervals() {
          clearInterval(cordovaCheckInterval);
          clearTimeout(buildTimeout);
          clearTimeout(errorTimeout);
          clearTimeout(messageTimeout);
          clearTimeout(killTimeout);
        }

        var args = _this4.prepareArguments();

        _this4.log.info(`running "meteor ${args.join(' ')}"... this might take a while`); // Lets spawn meteor.


        var child = (0, _crossSpawn.default)('meteor', args, {
          env: Object.assign({
            METEOR_PRETTY_OUTPUT: 0,
            METEOR_NO_RELEASE_CHECK: 1
          }, process.env),
          cwd: _this4.$.env.paths.meteorApp.root
        }, {
          shell: true
        }); // Kills the currently running meteor command.

        function kill() {
          sll('');
          child.kill('SIGKILL');

          if (self.$.env.os.isWindows) {
            windowsKill(child.pid);
          }
        }

        function exit() {
          killTimeout = setTimeout(function () {
            clearTimeoutsAndIntervals();
            desiredExit = true;
            kill();
            resolve();
          }, 500);
        }

        function copyBuild() {
          self.copyBuild().then(function () {
            exit();
          }).catch(function () {
            clearTimeoutsAndIntervals();
            kill();
            writeLog();
            reject('copy');
          });
        }

        cordovaCheckInterval = setInterval(function () {
          // Check if we already have cordova-build ready.
          if (_this4.isCordovaBuildReady()) {
            // If so, then exit immediately.
            if (_this4.indexHTMLstrategy === _this4.indexHTMLStrategies.INDEX_FROM_CORDOVA_BUILD) {
              copyBuild();
            }
          }
        }, 1000);
        child.stderr.on('data', function (chunk) {
          var line = chunk.toString('UTF-8');
          log += `${line}\n`;

          if (errorTimeout) {
            clearTimeout(errorTimeout);
          } // Do not exit if this is the warning for using --production.
          // Output exceeds -> https://github.com/meteor/meteor/issues/8592


          if (!~line.indexOf('--production') && !~line.indexOf('Output exceeds ')) {
            // We will exit 1s after last error in stderr.
            errorTimeout = setTimeout(function () {
              clearTimeoutsAndIntervals();
              kill();
              writeLog();
              reject('error');
            }, 1000);
          }
        });
        child.stdout.on('data', function (chunk) {
          var line = chunk.toString('UTF-8');

          if (!desiredExit && line.trim().replace(/[\n\r\t\v\f]+/gm, '') !== '') {
            var linesToDisplay = line.trim().split('\n\r'); // Only display last line from the chunk.

            var sanitizedLine = linesToDisplay.pop().replace(/[\n\r\t\v\f]+/gm, '');
            sll(sanitizedLine);
          }

          log += `${line}\n`;

          if (~line.indexOf('after_platform_add')) {
            sll('');

            _this4.log.info('done... 10%');
          }

          if (~line.indexOf('Local package version')) {
            if (messageTimeout) {
              clearTimeout(messageTimeout);
            }

            messageTimeout = setTimeout(function () {
              sll('');

              _this4.log.info('building in progress...');
            }, 1500);
          }

          if (~line.indexOf('Preparing Cordova project')) {
            sll('');

            _this4.log.info('done... 60%');
          }

          if (~line.indexOf('Can\'t listen on port')) {
            portProblem = true;
          }

          if (~line.indexOf('Your application has errors')) {
            if (errorTimeout) {
              clearTimeout(errorTimeout);
            }

            errorTimeout = setTimeout(function () {
              clearTimeoutsAndIntervals();
              kill();
              writeLog();
              reject('errorInApp');
            }, 1000);
          }

          if (~line.indexOf('App running at')) {
            copyBuild();
          }
        }); // When Meteor exits

        child.on('exit', function () {
          sll('');
          clearTimeoutsAndIntervals();

          if (!desiredExit) {
            writeLog();

            if (portProblem) {
              reject('port');
            } else {
              reject('exit');
            }
          }
        });
        buildTimeout = setTimeout(function () {
          kill();
          writeLog();
          reject('timeout');
        }, _this4.$.env.options.buildTimeout ? _this4.$.env.options.buildTimeout * 1000 : 600000);
      });
    }
    /**
     * Replaces the DDP url that was used originally when Meteor was building the client.
     * @param {string} indexHtml - path to index.html from the client
     */

  }, {
    key: "updateDdpUrl",
    value: function updateDdpUrl(indexHtml) {
      var content;
      var runtimeConfig;

      try {
        content = _fs.default.readFileSync(indexHtml, 'UTF-8');
      } catch (e) {
        this.log.error(`error loading index.html file: ${e.message}`);
        process.exit(1);
      }

      if (!this.matcher.test(content)) {
        this.log.error('could not find runtime config in index file');
        process.exit(1);
      }

      try {
        var matches = content.match(this.matcher);
        runtimeConfig = JSON.parse(decodeURIComponent(matches[1]));
      } catch (e) {
        this.log.error('could not find runtime config in index file');
        process.exit(1);
      }

      if (this.$.env.options.ddpUrl.substr(-1, 1) !== '/') {
        this.$.env.options.ddpUrl += '/';
      }

      runtimeConfig.ROOT_URL = this.$.env.options.ddpUrl;
      runtimeConfig.DDP_DEFAULT_CONNECTION_URL = this.$.env.options.ddpUrl;
      content = content.replace(this.replacer, `$1"${encodeURIComponent(JSON.stringify(runtimeConfig))}"$3`);

      try {
        _fs.default.writeFileSync(indexHtml, content);
      } catch (e) {
        this.log.error(`error writing index.html file: ${e.message}`);
        process.exit(1);
      }

      this.log.info('successfully updated ddp string in the runtime config of a mobile build' + ` to ${this.$.env.options.ddpUrl}`);
    }
    /**
     * Prepares the arguments passed to `meteor` command.
     * @returns {string[]}
     */

  }, {
    key: "prepareArguments",
    value: function prepareArguments() {
      var args = ['run', '--verbose', `--mobile-server=${this.$.env.options.ddpUrl}`];

      if (this.$.env.isProductionBuild()) {
        args.push('--production');
      }

      args.push('-p');

      if (this.$.env.options.port) {
        args.push(this.$.env.options.port);
      } else {
        args.push('3080');
      }

      if (this.$.env.options.meteorSettings) {
        args.push('--settings', this.$.env.options.meteorSettings);
      }

      return args;
    }
    /**
     * Validates the mobile build and copies it into electron app.
     */

  }, {
    key: "copyBuild",
    value: function () {
      var _copyBuild = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee5() {
        var prefix, copyPathPostfix, indexHtml, cordovaBuild, cordovaBuildIndex, cordovaBuildProgramJson, programJson;
        return regeneratorRuntime.wrap(function _callee5$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                this.log.debug('clearing build dir');
                _context5.prev = 1;
                _context5.next = 4;
                return this.$.utils.rmWithRetries('-rf', this.$.env.paths.electronApp.meteorApp);

              case 4:
                _context5.next = 9;
                break;

              case 6:
                _context5.prev = 6;
                _context5.t0 = _context5["catch"](1);
                throw new Error(_context5.t0);

              case 9:
                prefix = 'cordovaBuild';
                copyPathPostfix = '';

                if (!(this.indexHTMLstrategy === this.indexHTMLStrategies.INDEX_FROM_RUNNING_SERVER)) {
                  _context5.next = 27;
                  break;
                }

                prefix = 'webCordova';
                copyPathPostfix = `${_path.default.sep}*`;
                _context5.prev = 14;

                _fs.default.mkdir(this.$.env.paths.electronApp.meteorApp);

                _context5.next = 18;
                return this.acquireIndex();

              case 18:
                indexHtml = _context5.sent;

                _fs.default.writeFileSync(this.$.env.paths.electronApp.meteorAppIndex, indexHtml);

                this.log.info('successfully downloaded index.html from running meteor app');
                _context5.next = 27;
                break;

              case 23:
                _context5.prev = 23;
                _context5.t1 = _context5["catch"](14);
                this.log.error('error while trying to download index.html for web.cordova, ' + 'be sure that you are running a mobile target or with' + ' --mobile-server: ', _context5.t1);
                throw _context5.t1;

              case 27:
                cordovaBuild = this.$.env.paths.meteorApp[prefix];
                cordovaBuildIndex = this.$.env.paths.meteorApp.cordovaBuildIndex;
                cordovaBuildProgramJson = this.$.env.paths.meteorApp[`${prefix}ProgramJson`];

                if (this.$.utils.exists(cordovaBuild)) {
                  _context5.next = 34;
                  break;
                }

                this.log.error(`no mobile build found at ${cordovaBuild}`);
                this.log.error('are you sure you did run meteor with --mobile-server?');
                throw new Error('required file not present');

              case 34:
                if (this.$.utils.exists(cordovaBuildProgramJson)) {
                  _context5.next = 38;
                  break;
                }

                this.log.error('no program.json found in mobile build found at ' + `${cordovaBuild}`);
                this.log.error('are you sure you did run meteor with --mobile-server?');
                throw new Error('required file not present');

              case 38:
                if (!(this.indexHTMLstrategy !== this.indexHTMLStrategies.INDEX_FROM_RUNNING_SERVER)) {
                  _context5.next = 43;
                  break;
                }

                if (this.$.utils.exists(cordovaBuildIndex)) {
                  _context5.next = 43;
                  break;
                }

                this.log.error('no index.html found in cordova build found at ' + `${cordovaBuild}`);
                this.log.error('are you sure you did run meteor with --mobile-server?');
                throw new Error('required file not present');

              case 43:
                this.log.verbose('copying mobile build');

                _shelljs.default.cp('-R', `${cordovaBuild}${copyPathPostfix}`, this.$.env.paths.electronApp.meteorApp); // Because of various permission problems here we try to clear te path by clearing
                // all possible restrictions.


                _shelljs.default.chmod('-R', '777', this.$.env.paths.electronApp.meteorApp);

                if (this.$.env.os.isWindows) {
                  _shelljs.default.exec(`attrib -r ${this.$.env.paths.electronApp.meteorApp}${_path.default.sep}*.* /s`);
                }

                if (!(this.indexHTMLstrategy === this.indexHTMLStrategies.INDEX_FROM_RUNNING_SERVER)) {
                  _context5.next = 60;
                  break;
                }

                _context5.prev = 48;
                _context5.next = 51;
                return this.acquireManifest();

              case 51:
                programJson = _context5.sent;

                _fs.default.writeFileSync(this.$.env.paths.electronApp.meteorAppProgramJson, JSON.stringify(programJson, null, 4));

                this.log.info('successfully downloaded manifest.json from running meteor app');
                _context5.next = 60;
                break;

              case 56:
                _context5.prev = 56;
                _context5.t2 = _context5["catch"](48);
                this.log.error('error while trying to download manifest.json for web.cordova,' + ' be sure that you are running a mobile target or with' + ' --mobile-server: ', _context5.t2);
                throw _context5.t2;

              case 60:
                this.log.info('mobile build copied to electron app');
                this.log.debug('copy cordova.js to meteor build');

                _shelljs.default.cp(join(__dirname, '..', 'skeleton', 'cordova.js'), this.$.env.paths.electronApp.meteorApp);

              case 63:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee5, this, [[1, 6], [14, 23], [48, 56]]);
      }));

      return function copyBuild() {
        return _copyBuild.apply(this, arguments);
      };
    }()
    /**
     * Injects Meteor.isDesktop
     */

  }, {
    key: "injectIsDesktop",
    value: function injectIsDesktop() {
      var _this5 = this;

      this.log.info('injecting isDesktop');
      var manifestJsonPath = this.$.env.paths.meteorApp.cordovaBuildProgramJson;

      if (this.indexHTMLstrategy === this.indexHTMLStrategies.INDEX_FROM_RUNNING_SERVER) {
        manifestJsonPath = this.$.env.paths.meteorApp.webCordovaProgramJson;
      }

      try {
        var _JSON$parse = JSON.parse(_fs.default.readFileSync(manifestJsonPath, 'UTF-8')),
            manifest = _JSON$parse.manifest;

        var injected = false;
        var injectedStartupDidComplete = false;
        var result = null; // We will search in every .js file in the manifest.
        // We could probably detect whether this is a dev or production build and only search in
        // the correct files, but for now this should be fine.

        manifest.forEach(function (file) {
          var fileContents; // Hacky way of setting isDesktop.

          if (file.type === 'js') {
            fileContents = _fs.default.readFileSync(join(_this5.$.env.paths.electronApp.meteorApp, file.path), 'UTF-8');
            result = _this5.injector.processFileContents(fileContents);
            var _result = result;
            fileContents = _result.fileContents;
            injectedStartupDidComplete = result.injectedStartupDidComplete ? true : injectedStartupDidComplete;
            injected = result.injected ? true : injected;

            _fs.default.writeFileSync(join(_this5.$.env.paths.electronApp.meteorApp, file.path), fileContents);
          }
        });

        if (!injected) {
          this.log.error('error injecting isDesktop global var.');
          process.exit(1);
        }

        if (!injectedStartupDidComplete) {
          this.log.error('error injecting isDesktop for startupDidComplete');
          process.exit(1);
        }
      } catch (e) {
        this.log.error('error occurred while injecting isDesktop: ', e);
        process.exit(1);
      }

      this.log.info('injected successfully');
    }
    /**
     * Builds, modifies and copies the meteor app to electron app.
     */

  }, {
    key: "build",
    value: function () {
      var _build = _asyncToGenerator(
      /*#__PURE__*/
      regeneratorRuntime.mark(function _callee6() {
        return regeneratorRuntime.wrap(function _callee6$(_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                this.log.info('checking for any mobile platform');
                _context6.prev = 1;
                _context6.next = 4;
                return this.checkPreconditions();

              case 4:
                _context6.next = 10;
                break;

              case 6:
                _context6.prev = 6;
                _context6.t0 = _context6["catch"](1);
                this.log.error('error occurred during checking preconditions: ', _context6.t0);
                process.exit(1);

              case 10:
                this.log.info('building meteor app');

                if (this.$.env.options.skipMobileBuild) {
                  _context6.next = 41;
                  break;
                }

                _context6.prev = 12;
                _context6.next = 15;
                return this.buildMobileTarget();

              case 15:
                _context6.next = 39;
                break;

              case 17:
                _context6.prev = 17;
                _context6.t1 = _context6["catch"](12);
                _context6.t2 = _context6.t1;
                _context6.next = _context6.t2 === 'timeout' ? 22 : _context6.t2 === 'error' ? 24 : _context6.t2 === 'errorInApp' ? 26 : _context6.t2 === 'port' ? 28 : _context6.t2 === 'exit' ? 30 : _context6.t2 === 'copy' ? 32 : 34;
                break;

              case 22:
                this.log.error('timeout while building, log has been written to meteor.log');
                return _context6.abrupt("break", 35);

              case 24:
                this.log.error('some errors were reported during build, check meteor.log for more' + ' info');
                return _context6.abrupt("break", 35);

              case 26:
                this.log.error('your meteor app has errors - look into meteor.log for more' + ' info');
                return _context6.abrupt("break", 35);

              case 28:
                this.log.error('your port 3080 is currently used (you probably have this or other ' + 'meteor project running?), use `-t` or `--meteor-port` to use ' + 'different port while building');
                return _context6.abrupt("break", 35);

              case 30:
                this.log.error('meteor cmd exited unexpectedly, log has been written to meteor.log');
                return _context6.abrupt("break", 35);

              case 32:
                this.log.error('error encountered when copying the build');
                return _context6.abrupt("break", 35);

              case 34:
                this.log.error('error occurred during building mobile target', _context6.t1);

              case 35:
                if (!this.mobilePlatform) {
                  _context6.next = 38;
                  break;
                }

                _context6.next = 38;
                return this.removeMobilePlatform(this.mobilePlatform);

              case 38:
                process.exit(1);

              case 39:
                _context6.next = 50;
                break;

              case 41:
                this.indexHTMLstrategy = this.chooseStrategy();
                _context6.prev = 42;
                _context6.next = 45;
                return this.copyBuild();

              case 45:
                _context6.next = 50;
                break;

              case 47:
                _context6.prev = 47;
                _context6.t3 = _context6["catch"](42);
                process.exit(1);

              case 50:
                this.injectIsDesktop();
                this.changeDdpUrl();
                _context6.prev = 52;
                _context6.next = 55;
                return this.packToAsar();

              case 55:
                _context6.next = 61;
                break;

              case 57:
                _context6.prev = 57;
                _context6.t4 = _context6["catch"](52);
                this.log.error('error while packing meteor app to asar');
                process.exit(1);

              case 61:
                this.log.info('meteor build finished');

                if (!this.mobilePlatform) {
                  _context6.next = 65;
                  break;
                }

                _context6.next = 65;
                return this.removeMobilePlatform(this.mobilePlatform);

              case 65:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee6, this, [[1, 6], [12, 17], [42, 47], [52, 57]]);
      }));

      return function build() {
        return _build.apply(this, arguments);
      };
    }()
  }, {
    key: "changeDdpUrl",
    value: function changeDdpUrl() {
      if (this.$.env.options.ddpUrl !== null) {
        try {
          this.updateDdpUrl(this.$.env.paths.electronApp.meteorAppIndex);
        } catch (e) {
          this.log.error(`error while trying to change the ddp url: ${e.message}`);
        }
      }
    }
  }, {
    key: "packToAsar",
    value: function packToAsar() {
      var _this6 = this;

      this.log.info('packing meteor app to asar archive');
      return new Promise(function (resolve, reject) {
        return _asar.default.createPackage(_this6.$.env.paths.electronApp.meteorApp, _path.default.join(_this6.$.env.paths.electronApp.root, 'meteor.asar'), function () {
          // On Windows some files might still be blocked. Giving a tick for them to be
          // ready for deletion.
          setImmediate(function () {
            _this6.log.verbose('clearing meteor app after packing');

            _this6.$.utils.rmWithRetries('-rf', _this6.$.env.paths.electronApp.meteorApp).then(function () {
              resolve();
            }).catch(function (e) {
              reject(e);
            });
          });
        });
      });
    }
  }]);

  return MeteorApp;
}();

exports.default = MeteorApp;
